File: euler-java/src/net/projecteuler/level001/Problem0010.java
package net.projecteuler.level001;

import net.projecteuler.utils.Utils;

/*
 * The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
 *
 * Find the sum of all the primes below two million.
 */
public class Problem0010 {

  public static void main(String[] args) {

    Utils.time(Problem0010::solution);

  }

  public static void solution() {
    int limit = 2_000_000;

    boolean[] checked = new boolean[limit];

    long sum = 2;

    for (int i = 3; i < limit; i = i + 2) {
      if (checked[i]) {
        continue;
      }
      boolean flag = true;
      for (int j = 2; j * j < i; j++) {
        if (i % j == 0) {
          flag = false;
          break;
        }
      }

      if (flag) {

        sum = sum + i;
        int acc = i;
        while (acc < limit) {
          checked[acc] = true;
          acc += i;
        }
        // System.out.println(i);

      }

    }

    System.out.println(sum);
  }

}

File: euler-scala/src/net/projecteuler/level001/Problem0010.scala
package net.projecteuler.level001

import net.projecteuler.utils.{Prime, Stopwatch}

object Problem0010 {

  def main(args: Array[String]): Unit = {


    Stopwatch.elapsed(sieve(2000000))

  }


  def sieve(n: Int) = {

    val sieve = Array.ofDim[Boolean](n + 1)


    var sum = 2L

    var i = 3
    while(i <= n) {

      if(!sieve(i) && Prime.isPrime(i)) {
        sum += i

        var j = i + i
        while(j <= n) {
          sieve(j) = true
          j += i
        }
      }

      i += 2

    }


    println(sum)

  }

}
